src/Data/List/Relation/Unary/All/Properties.agda

```diff
@@ -419,14 +421,31 @@ module _ (S₁ : Setoid a ℓ₁) (S₂ : Setoid b ℓ₂) where
                           All P (cartesianProductWith f xs ys)
   cartesianProductWith⁺ f []       ys pres = []
   cartesianProductWith⁺ f (x ∷ xs) ys pres = ++⁺
-    (map⁺ (All.tabulateₛ S₂ (pres (here (Setoid.refl S₁)))))
+    (map⁺ (All.tabulateₛ S₂ (pres (here ≈₁-refl))))
     (cartesianProductWith⁺ f xs ys (pres ∘ there))
 
   cartesianProduct⁺ : ∀ xs ys →
                       (∀ {x y} → x ∈₁ xs → y ∈₂ ys → P (x , y)) →
                       All P (cartesianProduct xs ys)
   cartesianProduct⁺ = cartesianProductWith⁺ _,_
 
+  cartesianProductWith⁻ : ∀ f →
+                          f Preserves₂ _≈₁_ ⟶ _≈₂_ ⟶ _≡_ →
+                          ∀ xs ys →
+                          All P (cartesianProductWith f xs ys) →
+                          (∀ {x y} → x ∈₁ xs → y ∈₂ ys → P (f x y))
+  cartesianProductWith⁻ {P = P} f pres (x′ ∷ xs) ys Ps {y = y} x∈x′∷xs y∈ys
+    with x∈x′∷xs   | ++⁻ (map (f x′) ys) Ps
```

Comment:
I think it would be nicer the use plain pattern-matching for the ` x∈x′∷xs` argument and then dispatch on `with`.

Also: is there a way to use equational reasoning instead of `rewrite`? This proof is far from limpid.

Reply:
And rather than use `with`, as the result of that computation has a single-constructor type, it can be irrefutably matched with a pattern-matching `let` instead, or even the new `using` syntax?

Reply:
But, if the issue is one of *laziness*, namely that different projections are used in each branch, then better to use `proj₁`/`proj₂` instead?